services/src/main/java/org/keycloak/protocol/oidc/DefaultTokenExchangeProvider.java

```diff
@@ -335,6 +338,24 @@ protected Response exchangeClientToClient(UserModel targetUser, UserSessionModel
         }
 
         String scope = formParams.getFirst(OAuth2Constants.SCOPE);
+        if (token != null && token.getScope() != null && scope == null) {
+            scope = token.getScope();
```

Comment:
You are right. I have added it.

Reply:
Thanks. Now we make sure the resulting scopes are indeed valid for the target audience.

However, what will happen if a scope requires user consent but the consent was only granted for the original client (to which the subject token was issued for)? In this case, the scope will be granted even though the scope was not explicitly granted by the user during consent to the requesting client.

Not sure if it is an issue because we can assume that the client making the token exchange should inherit the consent previously granted so that it can seamlessly have access to the scope to invoke some other service. But that should only be true for confidential clients, I would say.

I can see two scenarios:

* Public client, the subject token is a token issue to this client, and the audience is some other client
* Confidential client (e.g.: backend), the subject token is from another client, and the audience is some other client

In the first case, the consent was granted to the public client so it is OK to also grant the scope to have access to the target audience.

But in the second case, the consent was not granted to the confidential client but to the client that the subject token was issued for.

Am I overthinking this?

Reply:
We believe that token exchange should not ask for user consent. Token exchange is a no browser flow and a user can not accept consent.
Now Keycloak ask for targetClient consent.

I believe that I should not make any change in consent in this PR. We should discuss consent in token exchange in the dedicated Token Exchange Breakout Session.
We believe that with this PR and a consent solution, OAuth 2.0 Token exchange will be production ready.